content/blog/2022/2022-01-16-spring-boot-null-safety-annotations.md

```diff
@@ -0,0 +1,260 @@
+
+We can express our intention using the `@Nullable` annotation on the field. This will tell the IDE that the field can be null in some cases, so no need to trigger an alarm. As the [JavaDoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/lang/Nullable.html) suggests:
+> Can be used in association with `@NonNullApi` or `@NonNullFields` to override the default non-nullable semantic to nullable.
+
+Similar to `NonNull`, the `Nullable` annotation can be applied to the method, parameter, or field level. 
+
+We can now mark the `pastEmployment` field as nullable:
+
+```java
+package io.reflectoring.nullsafety;
+
+// imports
+
+class Employee {
+
+  @Nullable
+  String pastEmployment;
+
+  @Nullable String getPastEmployment() {
+    return pastEmployment;
+  }
+
+  //...
+}  
+```
+Here is a quick summary for `@Nullable`:
+
+| Annotated element | Effect                                      |
+| ----------------- | ------------------------------------------- |
+| field             | Indicates that the field can be null        |
+| parameter         | Indicates that the parameter can be null    |
+| method            | Indicates that the method can return null   |
+| package           | Not Applicable                              |
+
+## Automated Build Checks
+
+So far, we are discussing how modern IDEs make it easier to write null-safe code. However, if we want to have some automated code checks in our build pipeline, that's also doable to some extent.
+
+[SpotBugs](https://spotbugs.github.io/) (the reincarnation of the famous but abandoned [FindBugs](http://findbugs.sourceforge.net/) project) offers a Maven/Gradle plugin that can detect code smells due to nullability. Let's see how we can use it.
+
+For a Maven project, we need to update the `pom.xml` to add the [SpotBugs Maven Plugin](https://spotbugs.readthedocs.io/en/latest/maven.html):
+
+```xml
+<plugin>
+  <groupId>com.github.spotbugs</groupId>
+  <artifactId>spotbugs-maven-plugin</artifactId>
+  <version>4.5.2.0</version>
+  <dependencies>
+    <!-- overwrite dependency on spotbugs if you want to specify the version of spotbugs -->
+    <dependency>
+      <groupId>com.github.spotbugs</groupId>
+      <artifactId>spotbugs</artifactId>
+      <version>4.5.3</version>
+    </dependency>
+  </dependencies>
+</plugin>
+```
+After building the project, we can use the following goals from this plugin:
+
+* the `spotbugs` goal analyzes the target project.
+* the `check` goal runs the `spotbugs` goal and makes the build fail if it finds any bugs.
+
+If you use Gradle instead of Maven, you can configure the [SpotBugs Gradle Plugin](https://spotbugs.readthedocs.io/en/latest/gradle.html) in your `build.gradle` file:
+
+```groovy
+dependencies {
+  spotbugsPlugins 'com.h3xstream.findsecbugs:findsecbugs-plugin:1.11.0'
+}
+
+spotbugs {
+  toolVersion = '4.5.3'
+}
+```
+Once the project is updated, we can run the check using the `gradle check` command.
```

Comment:
Last question before publishing. 

When I read this section about SpotBugs, I'm asking myself how this is connected to the Null annotations. Does SpotBugs evaluate these annotations? Or does SpotBugs have some other nullability checks? Would be good to explain that in the text. If SpotBugs doesn't evaluate the annotations, we should remove the section again, because this article is about the annotations.

Reply:
Hi Tom, thanks, I have slightly modified that section. Please have a look. SpotBugs could be another potential topic for us :)